src/node/txreconciliation.cpp

```diff
@@ -142,15 +347,131 @@ class TxReconciliationTracker::Impl
+        if (OUTBOUND_FANOUT_DESTINATIONS > outbounds_fanout_tx_relay) {
+            // This may only happen in testing, but make sure we don't have a target size bigger than our registered outbounds count
+            outbounds_target_size = std::min(OUTBOUND_FANOUT_DESTINATIONS - outbounds_fanout_tx_relay, m_states.size() - m_inbounds_count);
+        }
+
+        // Since we use the fraction for inbound peers, we first need to compute the total number of inbound targets.
+        const double inbound_targets = (inbounds_fanout_tx_relay + m_inbounds_count) * INBOUND_FANOUT_DESTINATIONS_FRACTION;
+        double n = std::max(inbound_targets - inbounds_fanout_tx_relay, 0.0);
+
+        // Being this a fraction, we need to round it either up or down. We do this deterministically at random based on the
+        // transaction we are picking the peers for.
+        CSipHasher deterministic_randomizer_in{m_deterministic_randomizer};
+        deterministic_randomizer_in.Write(wtxid.ToUint256());
+        CSipHasher deterministic_randomizer_out{deterministic_randomizer_in};
+        const size_t inbounds_target_size = ((deterministic_randomizer_in.Finalize() & 0xFFFFFFFF) + uint64_t(n * 0x100000000)) >> 32;
+
+        // Pick all reconciliation registered peers and assign them a deterministically random value based on their peer id
+        // Also, split peers in inbounds/outbounds
+        std::vector<std::pair<uint64_t, NodeId>> weighted_inbounds, weighed_outbounds;
+        weighted_inbounds.reserve(m_inbounds_count);
+        Assume(m_states.size() >= m_inbounds_count);
+        weighed_outbounds.reserve(m_states.size() - m_inbounds_count);
+
+        auto assign_key = [](const TxReconciliationState& peer_state, NodeId node_id, CSipHasher randomizer, std::vector<std::pair<uint64_t, NodeId>>& weighted_peers) {
+            uint64_t hash_key = randomizer.Write(node_id).Finalize();
+            weighted_peers.emplace_back(hash_key, node_id);
+        };
+        for (const auto& [node_id, op_peer_state]: m_states) {
+            const auto peer_state = std::get_if<TxReconciliationState>(&op_peer_state);
+            if (peer_state) {
+                if (peer_state->m_we_initiate) {
+                    assign_key(*peer_state, node_id, deterministic_randomizer_out, weighed_outbounds);
+                } else {
+                    assign_key(*peer_state, node_id, deterministic_randomizer_in, weighted_inbounds);
+                }
+
+            }
+        }
+
+        // Sort the peers based on their assigned random value, extract the node_ids and trim the collections to size
+        std::vector<NodeId> fanout_targets(inbounds_target_size + outbounds_target_size);
+        auto collect_fanout_targets = [&](std::vector<std::pair<uint64_t, NodeId>> weighted_peers, const size_t target_size) {
+            // Make sure we never select more targets than we can
+            Assert(outbounds_target_size <= weighed_outbounds.size());
```

Comment:
Fuzzed the class for a bit (fuzz test here: https://github.com/marcofleon/bitcoin/commits/erlay30116-fuzztest/) and this assertion failed.

I think having a preregistered peer and one inbound peer in `m_states` causes `outbound_target_size` to be 1 while `weighed_outbounds.size()` is 0 because `assign_key` is never called.

The equivalent unit test (which fails when I run it) would be something like
```c++
CSipHasher hasher(0x0706050403020100ULL, 0x0F0E0D0C0B0A0908ULL);
TxReconciliationTracker tracker(TXRECONCILIATION_VERSION, hasher);
NodeId peer_id0 = 0;
NodeId peer_id1 = 1;
FastRandomContext frc{/*fDeterministic=*/true};
std::vector<NodeId> fanout_targets;

tracker.PreRegisterPeer(peer_id0);
tracker.PreRegisterPeer(peer_id1);
BOOST_REQUIRE_EQUAL(tracker.RegisterPeer(peer_id1, /*is_peer_inbound=*/true, 1, 1), ReconciliationRegisterResult::SUCCESS);
fanout_targets = tracker.GetFanoutTargets(Wtxid::FromUint256(frc.rand256()), /*inbounds_fanout_tx_relay=*/0, /*outbounds_fanout_tx_relay=*/0);
```

Maybe it makes sense to have separate maps for preregistered vs registered peers?

Reply:
Great fuzzing!

ahhhh the sloppy part is `m_states.size() - m_inbounds_count;` , it ignores the presence of preregistered peers. 

> Maybe it makes sense to have separate maps for preregistered vs registered peers?

I'd rather have `m_outbounds_count` cached value (all similar to `m_inbounds_count`), but no big deal.